scaaml/capture/aes_capture_util.py

```diff
@@ -0,0 +1,225 @@
+      resume_kti: An iterator of (key, text) pairs to capture traces with.
+      dataset: The dataset being created (for writing the actual traces).
+      examples_per_shard: How many captures are in a single shard.
+      plaintexts: How many plaintexts per a key value should be generated.
+      repetitions: How many traces should be captured with a single
+        (key, text) pair.
+      chip_id: Unique id of the piece of hardware.
+      purpose: Purpose of this dataset, in ['train', 'holdout'].
+    """
+    # Load the measurement points using resume_kti and finish the campaign
+    assert len(
+        resume_kti) % examples_per_shard == 0  # Whole shards are loaded.
+    # A part is the id of the shard of a single key, part is in [0, 10].
+    part_number = 0
+    group_number = 0  # a group is 1 full rotation of the bytes (0 - 255)
+    trace_number = 0  # Enumerating the for-loop.
+    for key, text in tqdm(resume_kti):
+        while True:  # Make sure to capture the trace before moving to the next.
+            # Try to capture trace.
+            cur_key = bytearray(key)
+            cur_pt = bytearray(text)
+            trace = cw.capture_trace(scope=cwscope,
+                                     target=target,
+                                     plaintext=cur_pt,
+                                     key=cur_key)
+            if trace:
+                # Captures 0, 1 ,..., examples_per_shard-1 form a single shard.
+                if trace_number % examples_per_shard == 0:
+                    if trace_number % (plaintexts * repetitions) == 0:
+                        # One key has been processed, reset the part_number.
+                        part_number = 0
+                    # Initialize a new shard.
+                    dataset.new_shard(
+                        key=cur_key,
+                        part=part_number,
+                        group=group_number,
+                        split=purpose,
+                        chip_id=chip_id,
+                    )
+                    part_number += 1
+                    if (trace_number > 0 and trace_number %
+                        (256 * plaintexts * repetitions) == 0):
+                        group_number += 1
+                assert trace.textin == cur_pt
+                dataset.write_example(attack_points={
+                    'sub_bytes_in':
+                    AES_SBOX.sb_in(key=cur_key, text=cur_pt),
+                    'sub_bytes_out':
+                    AES_SBOX.sb_out(key=cur_key, text=cur_pt),
+                    'key':
+                    cur_key,
+                },
+                                      measurement={
+                                          "trace": trace.wave,
+                                      })
+                break  # from capturing the trace (while True)
+        # Increment the enumeration of the traces.
+        trace_number += 1
+    # Make sure we close the last shard
+    dataset.close_shard()
```

Comment:
capturing may raise all sorts of exceptions (hardware failing, etc.) especially when it takes several hundreds hours to capture.
Don't we want something like a `contextmanager` to handle the resources nicely and make sure we close all the handles?
For example, when using Chipwhisperer scope, any failure at disconnecting/closing the resources (`cwscope.dis()` and `target.dis()`) will result in the hardware lock not being released. For there, one needs to physically replug the USB cable to access again the hardware. 

All functions/classes interacting with hardware should make sure that an uncaught exception will leave the hardware in a stable state where we can simply relaunch/resume the capture without physical intervention.